File: QLNS/src/main/java/QuanLy/QLNS/Entity/DanhGia.java
package QuanLy.QLNS.Entity;

import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import java.time.LocalDate;
import java.time.LocalDateTime;

@Entity
@Table(name = "danh_gia")
@Data
@NoArgsConstructor
@AllArgsConstructor
public class DanhGia {
    
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long danhgia_id;
    
    @ManyToOne
    @JoinColumn(name = "nhanvien_id", nullable = false)
    private NhanVien nhanVien;
    
    @ManyToOne
    @JoinColumn(name = "nguoi_danh_gia_id", nullable = false)
    private NhanVien nguoiDanhGia;
    
    @Column(name = "ky_danh_gia", nullable = false)
    private LocalDate kyDanhGia; // Kỳ đánh giá (tháng/quý/năm)
    
    @Column(name = "loai_danh_gia", length = 50)
    private String loaiDanhGia; // THANG, QUY, NAM, THU_VIEC
    
    @Column(name = "diem_chuyen_mon", nullable = false)
    private Integer diemChuyenMon; // 1-10
    
    @Column(name = "diem_thai_do", nullable = false)
    private Integer diemThaiDo; // 1-10
    
    @Column(name = "diem_ky_nang", nullable = false)
    private Integer diemKyNang; // 1-10
    
    @Column(name = "diem_trung_binh")
    private Double diemTrungBinh;
    
    @Column(name = "xep_loai", length = 50)
    private String xepLoai; // XUAT_SAC, TOT, KHA, TRUNG_BINH, YEU
    
    @Column(name = "nhan_xet", columnDefinition = "TEXT")
    private String nhanXet;
    
    @Column(name = "muc_tieu_ke_tiep", columnDefinition = "TEXT")
    private String mucTieuKeTiep;
    
    @Column(name = "ngay_danh_gia")
    private LocalDateTime ngayDanhGia;
    
    @PrePersist
    protected void onCreate() {
        ngayDanhGia = LocalDateTime.now();
        // Tính điểm trung bình
        diemTrungBinh = (diemChuyenMon + diemThaiDo + diemKyNang) / 3.0;
        // Xếp loại tự động
        if (diemTrungBinh >= 9) xepLoai = "XUAT_SAC";
        else if (diemTrungBinh >= 8) xepLoai = "TOT";
        else if (diemTrungBinh >= 6.5) xepLoai = "KHA";
        else if (diemTrungBinh >= 5) xepLoai = "TRUNG_BINH";
        else xepLoai = "YEU";
    }
}

File: QLNS-frontend/src/api/danhgia.ts
import { keepPreviousData, useMutation, useQuery, useQueryClient } from '@tanstack/react-query'
import api from './client'

export type DanhGia = {
  danhgia_id: number
  nhanVien: { nhanvien_id: number; ho_ten?: string } | number
  nguoiDanhGia: { nhanvien_id: number; ho_ten?: string } | number
  kyDanhGia: string
  loaiDanhGia?: 'THANG' | 'QUY' | 'NAM' | 'THU_VIEC' | string
  diemChuyenMon: number
  diemThaiDo: number
  diemKyNang: number
  diemTrungBinh?: number
  xepLoai?: 'XUAT_SAC' | 'TOT' | 'KHA' | 'TRUNG_BINH' | 'YEU' | string
  nhanXet?: string
  mucTieuKeTiep?: string
  ngayDanhGia?: string
}

export type Page<T> = {
  content: T[]
  totalElements: number
  totalPages: number
  number: number
  size: number
}

export function useDanhGiaList(page: number, size: number) {
  return useQuery({
    queryKey: ['danhgia', { page, size }],
    queryFn: async () => {
      const res = await api.get<Page<DanhGia>>('/api/danhgia', { params: { page, size } })
      return res.data
    },
    placeholderData: keepPreviousData,
  })
}

export function useCreateDanhGia() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: async (body: Partial<DanhGia>) => {
      const res = await api.post<DanhGia>('/api/danhgia', body)
      return res.data
    },
    onSuccess: () => qc.invalidateQueries({ queryKey: ['danhgia'] }),
  })
}

export function useUpdateDanhGia() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: async ({ id, body }: { id: number; body: Partial<DanhGia> }) => {
      const res = await api.put<DanhGia>(`/api/danhgia/${id}`, body)
      return res.data
    },
    onSuccess: (_d, vars) => {
      qc.invalidateQueries({ queryKey: ['danhgia'] })
      qc.invalidateQueries({ queryKey: ['danhgia', vars.id] })
    },
  })
}

export function useDeleteDanhGia() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: async (id: number) => {
      await api.delete(`/api/danhgia/${id}`)
      return id
    },
    onSuccess: () => qc.invalidateQueries({ queryKey: ['danhgia'] }),
  })
}
